crates/red_knot_python_semantic/src/types/property_tests/arbitrary.rs

```
@@ -0,0 +1,306 @@
+            Ty::SubclassOfAny,
+            Ty::SubclassOfBuiltinClass("object"),
+            Ty::SubclassOfBuiltinClass("str"),
+            Ty::SubclassOfBuiltinClass("type"),
+            Ty::AbcClassLiteral("ABC"),
+            Ty::AbcClassLiteral("ABCMeta"),
+            Ty::SubclassOfAbcClass("ABC"),
+            Ty::SubclassOfAbcClass("ABCMeta"),
+            Ty::AlwaysTruthy,
+            Ty::AlwaysFalsy,
+            Ty::BuiltinsFunction("chr"),
+            Ty::BuiltinsFunction("ascii"),
+            Ty::BuiltinsBoundMethod {
+                class: "str",
+                method: "isascii",
+            },
+            Ty::BuiltinsBoundMethod {
+                class: "int",
+                method: "bit_length",
+            },
+        ])
+        .unwrap()
+        .clone()
+    }
+}
+
+/// A constraint that passes `t.is_fully_static()`.
+#[derive(Debug, Clone)]
+pub(crate) struct FullyStaticTy;
+
+impl ArbitraryRule for FullyStaticTy {
+    fn generate_singular_type(g: &mut Gen) -> Ty {
+        let int_lit = Ty::IntLiteral(*g.choose(&[-2, -1, 0, 1, 2]).unwrap());
+        let bool_lit = Ty::BooleanLiteral(bool::arbitrary(g));
+        g.choose(&[
+            Ty::Never,
+            Ty::BuiltinsFunction("chr"),
+            Ty::BuiltinsFunction("ascii"),
+            Ty::BuiltinsBoundMethod {
+                class: "str",
+                method: "isascii",
+            },
+            Ty::BuiltinsBoundMethod {
+                class: "int",
+                method: "bit_length",
+            },
+            int_lit,
+            bool_lit,
+            Ty::StringLiteral(""),
+            Ty::StringLiteral("a"),
+            Ty::LiteralString,
+            Ty::BytesLiteral(""),
+            Ty::BytesLiteral("\x00"),
+            Ty::KnownClassInstance(KnownClass::Object),
+            Ty::KnownClassInstance(KnownClass::Str),
+            Ty::KnownClassInstance(KnownClass::Int),
+            Ty::AlwaysFalsy,
+            Ty::AlwaysTruthy,
```

It looks to me like there are some fully-static types in the above list that are not included here, silently reducing our type coverage for tests using `FullyStaticTy`. (To pick just one example, `BuiltinClassLiteral('str')` is both fully-static and a singleton, but is not included either here or below.)

I think this is a significant drawback of this PR. Rather than having a single pool of candidate tests, with the actual implemention of `is_fully_static` and `is_singleton` deciding which tests fall into those categories, we now have to manually sort them. And while the new `ensure_rules` tests help validate that we don't include any type in a category to which it doesn't belong, there's nothing helping us validate that we aren't silently losing coverage by omitting types wrongly.

If we do decide to go ahead with this PR, I would at least prefer if we could not repeat ourselves. That is, have a single list of non-fully-static types, a single list of static types, and a single list of singleton types. `FullyStaticTy` would combine the latter two lists, `AnyTy` would combine all three, `SingletonTy` would use only the latter list.

---

First of all, I think I should explain that my initial intuition was off. I had assumed that using `is_fully_static` would make it very unlikely for valid tests—but it turns out that’s not really the case.

From my simulation, the probability that all three variables of `AnyTy` satisfy the `is_fully_static` premise is around 65%. That means, out of 100 test runs, about 65 actually exercise the property, which is higher than I expected.

The key point I was trying to focus on wasn’t performance—it was test validity. If the probability of satisfying a property’s premise approaches zero, the test becomes meaningless. This PR was based on that assumption, so the exact ratio of `AnyTy` to `FullyStaticTy` wasn’t my main concern.\
But as you've pointed out, `FullyStaticTy` doesn’t seem to be as effective as I initially thought.

In conclusion, I think it's totally fine to close this PR.

---

Here are a few more stats from the simulation, which might be helpful for future property test reviews:

- Probability that `t.is_singleton()` holds: \~9%
- Probability that `!t.is_fully_static()` holds: \~10%
- Probability that `s.is_equivalent_to(db, t) && t.is_equivalent_to(db, u)` holds: \~0.1%
- Probability that `s.is_gradual_equivalent_to(db, t)` holds: \~1%

These probabilities will likely decrease as we add more types to the random pool. But for now, it seems we can compensate by increasing the number of tests. If we move forward with similar PRs in the future, I think we should aim to eliminate premises with very low probability, not fully static.

---

Aha, I hadn't fully processed the fact that if we have a premise of low probability, that doesn't just impact performance of the test, it also impacts usefulness of the test; a premise of 10% probability means we need 10x the `QUICKCHECK_TESTS` value to get the same number of actual tests of the implication. I agree this is a stronger motivation for the PR, and could be worth the risk of missing coverage of some type that we fail to include in the right list.

I would consider going ahead with this PR if we implement the above suggestion to avoid repetition in the type lists (so each type is listed exactly once, making it easier to evaluate whether each type is in the right place). We could also consider eliminating `FullyStaticTy` and just separating `SingletonTy`, but I kind of think if we go down this route we may as well do it fully.

But I would also like to hear @sharkdp 's opinion, so maybe no need to do any updates here until he is back.

---

This PR looks great! It's something I've been meaning to implement as well. I kind of liked the generic and simple `premise => property` form, but with the syntactic change to the macro, this reads even better, I think.

> with the actual implemention of `is_fully_static` and `is_singleton` deciding which types fall into which categories, we now have to manually categorize the candidate types.

Maybe we can maintain that property? We could have a single pool of types and then categorize them into (static, `LazyLock`) specialized lists for usage in those dedicated `Arbitrary` implementations?

---

Sounds good! I’ll go ahead with the PR.

As sharkdp suggested, I think keeping a single pool and letting the `Arbitrary` implementations handle the categorization makes sense — it should make it easier to support more premises in the future.